src/System.Linq.Parallel/tests/Combinatorial/SourcesAndOperators.cs

```diff
@@ -0,0 +1,304 @@
+            yield return Label("Intersect-Left:" + label, (start, count, source) => other(start, count + count / 2).Intersect(source(start - count / 2, count + count / 2), new ModularCongruenceComparer(count * 2)));
+
+            yield return Label("Join-Right:" + label, (start, count, source) => source(0, count).Join(other(start, count), x => x, y => y - start, (x, y) => x + start, new ModularCongruenceComparer(count)));
+            yield return Label("Join-Left:" + label, (start, count, source) => other(0, count).Join(source(start, count), x => x, y => y - start, (x, y) => x + start, new ModularCongruenceComparer(count)));
+
+            yield return Label("Union-Right:" + label, (start, count, source) => source(start, count * 3 / 4).Union(other(start + count / 2, count / 2 + count % 2)));
+            yield return Label("Union-Left:" + label, (start, count, source) => other(start, count * 3 / 4).Union(source(start + count / 2, count / 2 + count % 2)));
+
+            // When both sources are unordered any element can be matched to any other, so a different check is required.
+            yield return Label("Zip-Unordered-Right:" + label, (start, count, source) => source(0, count).Zip(other(start * 2, count), (x, y) => x + start));
+            yield return Label("Zip-Unordered-Left:" + label, (start, count, source) => other(start * 2, count).Zip(source(0, count), (x, y) => y + start));
+        }
+
+        public static IEnumerable<object[]> BinaryUnorderedOperators()
+        {
+            LabelledOperation otherSource = UnorderedRangeSources().First();
+            foreach (LabelledOperation source in UnorderedRangeSources())
+            {
+                // Operations having multiple paths to check.
+                foreach (LabelledOperation operation in BinaryOperations(otherSource))
+                {
+                    yield return new object[] { source, operation };
+                }
+            }
+        }
+
+        public static IEnumerable<object[]> BinaryOperators()
+        {
+            LabelledOperation unordered = UnorderedRangeSources().First();
+            foreach (LabelledOperation source in RangeSources())
+            {
+                // Each binary can work differently, depending on which of the two source queries (or both) is ordered.
+
+                // For most, only the ordering of the first query is important
+                foreach (LabelledOperation operation in BinaryOperations(unordered).Where(op => !(op.ToString().StartsWith("Union") || op.ToString().StartsWith("Zip")) && op.ToString().Contains("Right")))
+                {
+                    yield return new object[] { source, operation };
+                }
+
+                // For Concat, since either one can be ordered the other side has to be tested
```

Comment:
I'm still waiting to hear if this is an actual bug or not, but yeah, I can add it.